models/policy_document.go

```diff
@@ -0,0 +1,88 @@
+package models
+
+import (
+	"encoding/json"
+	"fmt"
+)
+
+type PolicyDocument struct {
+	Version   string
+	Statement []statement
+}
+
+type statement struct {
+	Sid       string
+	Effect    string
+	Principal map[string]string
+	Action    []string
+}
+
+func NewPolicyDocument() PolicyDocument {
+	return PolicyDocument{
+		Version:   "2008-10-17",
+		Statement: []statement{},
+	}
+}
+
+func newStatement(accountID string) statement {
+	return statement{
+		Sid:    accountID,
+		Effect: "allow",
+		Principal: map[string]string{
+			"AWS": fmt.Sprintf("arn:aws:iam::%s:root", accountID),
+		},
+		Action: []string{
+			"ecr:GetDownloadUrlForLayer",
+			"ecr:BatchGetImage",
+			"ecr:BatchCheckLayerAvailability",
+		},
+	}
+}
+
+func (p *PolicyDocument) AddAWSAccountPrincipal(accountID string) bool {
+	permissionExists := false
+	for _, s := range p.Statement {
+		if s.Sid == accountID {
+			permissionExists = true
+			break
+		}
+	}
+
+	if permissionExists {
+		return false
+	}
+
+	p.Statement = append(p.Statement, newStatement(accountID))
+	return true
+}
+
+func (p *PolicyDocument) RemoveAWSAccountPrincipal(accountID string) bool {
+	temp := []statement{}
+	result := false
+
+	for _, s := range p.Statement {
+		if s.Sid == accountID {
+			result = true
+			continue
+		}
+
+		temp = append(temp, s)
+	}
+
+	p.Statement = temp
+	return result
+}
+
+func (p *PolicyDocument) RenderPolicyText() string {
```

Comment:
I know it's not a huge deal as-is, but I'm more worried about some dev seeing this in the future, thinking this is an acceptable pattern, and then using it somewhere where it actually does become an issue.
I think it would be wise to either:
* Change the return to `string, error`
* Remove this function and leave marshalling up to the caller

Reply:
Also, just in case you didn't already know - you can customize how objects get marshaled with `json.Marshal` by implementing the [Marshaler interface](http://choly.ca/post/go-json-marshalling/). 

I don't think this is a necessary or even a very useful option in this context, but figured it was a nice thing worth knowing.

Reply:
✅